### PR TITLE
Version 2.0

### PR DESCRIPTION
**Pull request by esimonetti**
Performance improvements (originally 3000 identical records would process in 159, now the same records process in 13 seconds or less)
Background queueing when more than 100 records
User notification after finishing, if it was an admin initiated action
Additional code improvements

### CODE REVIEW

File: src/custom/src/OutboundEmailsDeployer/OutboundEmailsDeployer.php [hunk 1]
```diff
@@ -314,38 +373,44 @@ protected function removeOrphanMailboxes($mapping)
...
-    protected function areOutboundMailboxesInSync($originalMailbox, $usersMailbox)
+    protected function areOutboundMailboxesInSync($originalMailbox, $usersMailboxDB)
     {
         $this->enforcePermissions();
 
-        if (!empty($originalMailbox->id) && !empty($usersMailbox->id)) {
+        if (!empty($originalMailbox->id) && !empty($usersMailboxDB->id)) {
+            // decode the pass for comparison, since we now compare against the database values
+            foreach ($this->fieldsToDecode as $fieldToDecode) {
+                $usersMailboxDB->$fieldToDecode = Blowfish::decode(Blowfish::getKey($originalMailbox->module_key), $usersMailboxDB->$fieldToDecode);
```
Comment (glevine) [inline src/custom/src/OutboundEmailsDeployer/OutboundEmailsDeployer.php]: `SugarBean` has this function...

```
public function decrypt_after_retrieve($value)
{
    if(empty($value)) return $value; // no need to decrypt empty
    return Blowfish::decode($this->getEncryptKey(), $value);
}
```

You just need to make populate a bean with the data from the row you retrieved, if `$usersMailboxDB` isn't already a bean. But I wouldn't use the method directly, unless you have to. `SugarBean` also has...

```
public function &__get($varname)
{
    if(isset($this->encfields[$varname])) {
        $this->$varname = $this->decrypt_after_retrieve($this->encfields[$varname]);
        return $this->$varname;
    }
    $var = null;
    return $var;
}
```

Simply by making `$usersMailboxDB` a bean, you will automatically get the decoded value of the field anytime you use the getter via `$usersMailboxDB->mail_smtppass`. You can convert an array from a database row into a bean with...

```
$row = [...];
$usersMailbox = BeanFactory::newBean('OutboundEmail');
$usersMailbox->populateFromRow($row, true); // I'm not 100% whether or not you'll need to pass `true`. I would need to investigate it further. But that's what SugarBean::retrieve() does.
```

Basically, you can still fetch data using your own query, but you can instantiate a bean so that you get all of the conveniences that come with beans.
